### PR TITLE
Remove singletons from audio stream

### PR DESCRIPTION
**Pull request by dzosz**
This change allows for creating more sound streams. We currently have only one dedicated to BG music. https://github.com/beyond-all-reason/spring/blob/414df1e8bb81535792fe9e77a6dcc0ae8c44e414/rts/System/Sound/ISoundChannels.h#L11
 
TODO rewrite SoundSource and AudioChannel and split samples from streams?

### CODE REVIEW

**Comment by lhog** [comment]
What is the use case for this one?

**Comment by sprunk** [comment]
> allows for creating more sound streams

Which would be very useful, examples:
* crossfade music tracks (requires two parallel streams for a few seconds when the old one is ending and the new one is picking up)
* dynamic music composed of layers that enable or disable based on circumstances. For example there's a "baseline" melody but if the game has escalated into berthas then there's also a kettle-drums overlay on top (= second parallel stream), and there's a badass electric guitar / sad flute if you're winning / losing (= third parallel stream).
* just having two parallel streams (for example music and rain).
* some unit and/or ambient sounds could be streams (for example "vrrrr" tank movement, or some sort of busy factory that makes with noise, or burning napalm) which requires multiple if there are multiple units. Currently you can only do this by repeating individual sounds which is bad (if they're long they can outlast the action they represent, if they're short they get repetitve) so no game really does that. Would also require streams to be playable 3D in-world (so might require a separate PR, idk).

**Comment by lhog** [comment]
@dzosz Is this one complete? I don't feel confident in this area of the engine, so will rely on your suggestion.

**Comment by dzosz** [comment]
I tested multiple streams using modified PlaySoundStream() interface - all good. While testing I found issues with stream queue (unrelated to this PR, currently unused as the queue is handled in lua) and reported them here: https://github.com/beyond-all-reason/spring/issues/951 . They should be dealt with when implementing cross-fading tracks.
Interface for dealing with multiple streams is still to be determined (depending on the features we want to support) but I'm happy to push this small refactoring out.

**Comment by sprunk** [comment]
Rebased and fixed merge conflict.

Pushed some trivial commits directly:
* 35c4eefc3427a50f0818dc3b795fe74ea64250db
* 7db9bbf7213da6d35e8a8192be8ebbfee0f38750
* 268a439e2908a8dc6f82ded337e6012c1f987e06
* bcc0f71b622f5ec479c84e76bce5a829c24cc18e

File: rts/System/Sound/OpenAL/MusicStream.cpp [hunk 1]
```diff
@@ -1,36 +1,19 @@
...
 
-// FIXME these parts of MusicStream are outside the class because every instance
-// of CSoundSource has a copy of MusicStream class but only one is ever active
-static std::variant<OggDecoder, Mp3Decoder> decoder;
-
-
-MusicStream::MusicStream(ALuint _source)
-	: source(_source)
+MusicStream::MusicStream()
+	: buffers{}
```
Comment (sprunk) [inline rts/System/Sound/OpenAL/MusicStream.cpp:15]: There's `std::fill` left in the constructor body, should be redundant with `buffers{}`

File: rts/System/Sound/OpenAL/SoundSource.cpp [hunk 2]
```diff
@@ -53,6 +54,15 @@ CSoundSource::CSoundSource()
 	curPlayingItem = {0,  0, 0,  0.0f, 0.0f};
 }
 
+CSoundSource::CSoundSource(CSoundSource&& src)
```
Comment (sprunk) [inline rts/System/Sound/OpenAL/SoundSource.cpp:57]: Can't this be `= default`?

File: rts/System/Sound/OpenAL/AudioChannel.h [hunk 3]
```diff
@@ -24,6 +26,7 @@ class AudioChannel : public IAudioChannel {
 	typedef std::pair<std::string, float> StreamQueueItem;
 
 public:
+	AudioChannel();
```
Comment (sprunk) [inline rts/System/Sound/OpenAL/AudioChannel.h]: This gets added in this commit and removed in the next one? Also the `= default` is in the `.cpp` file which is somewhat inconsistent with how it looks in other classes.

Comment (dzosz) [reply rts/System/Sound/OpenAL/AudioChannel.h]: The first idea was to cache `MusicStream` in channel but these changes were reverted.

File: rts/System/Sound/OpenAL/SoundSource.cpp (hunk 2 above)
Comment (dzosz) [reply rts/System/Sound/OpenAL/SoundSource.cpp:63]: implemented move operations in [8b8cc67](https://github.com/beyond-all-reason/spring/pull/857/commits/8b8cc670030cc299864c60a26e56c43562bb6e4a)

Comment (lhog) [reply rts/System/Sound/OpenAL/SoundSource.cpp:63]: Why `swap()` and not the move assignment operator and constructor?

Comment (dzosz) [reply rts/System/Sound/OpenAL/SoundSource.cpp:63]: > Why `swap()` and not the move assignment operator and constructor?

default move operator copies numeric `id` field which needs to be unique for each object instance

**Review by sprunk — APPROVED** [review]
I don't recall any problems that wouldn't be gone after squashing.

**Comment by dzosz** [comment]
I'm quite busy recently so I might not be work on potential issues after merge.

**Comment by sprunk** [comment]
@dzosz how's your availability?

**Comment by dzosz** [comment]
> @dzosz how's your availability?

I'm available. I'm looking into the changes that will be required to make use of multiple streams.

**Review by sprunk — APPROVED** [review]

**Comment by sprunk** [comment]
Rebased and cleaned up. Did a basic test (just run mp3 and ogg streams a couple of times) and found no issues. IMO mergeable @lostsquirrel1

**Pull request merged**